packages/babel-parser/src/parser/expression.js

```diff
@@ -2583,12 +2579,16 @@ export default class ExpressionParser extends LValParser {
       throw this.unexpected();
     }
 
+    const tokenIsKeyword = tokenKeywordOrIdentifierIsKeyword(type);
+
     if (liberal) {
       // If the current token is not used as a keyword, set its type to "tt.name".
       // This will prevent this.next() from throwing about unexpected escapes.
-      this.state.type = tt.name;
+      if (tokenIsKeyword) {
+        this.replaceToken(tt.name);
```

Comment:
Is `updateContext` needed when replacing a keyword with `tt.name`? (I don't think it is).
Because we could just to `this.state.type = tt.name` here (and maybe `this.state.exprAllowed = false`, even if I don't see why it would be needed) and keep the `prevType` handling all in the JSX plugin (to avoid needing the type error suppression comment).

Reply:
Uh, or maybe it would break this:
```js
foo.else
<X>2</X>/* 2//*/
```

Reply:
The `updateContext` is currently required by the jsx plugin, to disable JSX element forming in the following situations, where a keyword token was re-interpreted as an identifier token and thus it can not start a JSX element:

```js
x.case <jsx ...
x?.case <jsx ...
class C { case <jsx ...
```